compressor.go

```diff
@@ -32,6 +32,7 @@ type Compressor interface {
 	Name() string
 	Encode(data []byte) ([]byte, error)
 	Decode(data []byte) ([]byte, error)
+	DecodeSized(data []byte, size uint32) ([]byte, error)
```

Comment:
Could you please add doc comment describing the semantics of the method?

<hr />

We could also change this to be append-like, as follows:

```go
// DecodeSized decodes the encoded bytes and appends them to dst.
// decodeSize is the size of data after decompression.
DecodeSized(dst, encoded []byte, decodedSize int32) ([]byte, error)
```

It seems that would be more flexible, for example it would allow the framer to reuse buffers for decompression in a sync.Pool if needed. And since this will be in 2.0, we could adjust the other methods as well.

Reply:
I agree with this, it's a great suggestion 👍

Reply:
Then we should agree on the API.

We could adjust the current API to [this](https://github.com/worryg0d/gocql/commit/b4e0322e2af541fa0a3b7c467802e20cfa691ef6):
```go
type Compressor interface {
	Name() string
	Encode(dst, data []byte) ([]byte, error)
	Decode(dst, data []byte) ([]byte, error)

	// DecodeSized decodes the encoded bytes and appends them to dst.
	// decodeSize is the size of data after decompression.
	DecodeSized(dst, encoded []byte, decodedSize uint32) ([]byte, error)
}
```

As another solution, we could change the API to this:
```go
type Compressor interface {
	Name() string
	
	// EncodeWithSize encodes data bytes, appends the size of the data
	// to dst and appends encoded bytes.
	// It returns encoded data length.
	EncodeWithSize(dst, data []byte) (n int, err error)
	
	// DecodeWithSize reads size of uncompressed data from the encoded data,
	// decodes the encoded bytes and appends them to dst.
	// It returns decoded data length.
	DecodeWithSize(dst, encoded []byte) (n int, err error)

	// Encode encodes data bytes and appends them to dst.
	// It returns encoded data length.
	Encode(dst, data []byte) (n int, err error)

	// Decode decodes the encoded bytes and appends them to dst.
	// It returns decoded data length.
	Decode(dst, encoded []byte) (n int, err error)
}
```

The second provides distinguishing methods for:
* **frame body** encoding/decoding `EncodeWithSize` and `DecodeWithSize`
* **segment payload** processing `Encode` and `Decode`.

Reply:
To truly make it 100% "append-like" (examples: https://github.com/golang/go/issues/53693):

```
type Compressor interface {
	Name() string
	
	// AppendCompressedWithLength compresses src bytes, appends the length of the compressed bytes to dst and then appends the compressed bytes to dst.
        // It returns a new byte slice that is the result of the append operation.
	AppendCompressedWithLength(dst, src []byte) ([]byte, error)
	
	// AppendDecompressedWithLength reads the length of the decompressed bytes from src, decompresses bytes from src and appends the decompressed bytes to dst.
        // It returns a new byte slice that is the result of the append operation.
	AppendDecompressedWithLength(dst, src []byte) ([]byte, error)

	// AppendCompressed compresses src bytes and appends the compressed bytes to dst.
        // It returns a new byte slice that is the result of the append operation.
	AppendCompressed(dst, src []byte) ([]byte, error)

	// AppendDecompressed decompresses bytes from src and appends the decompressed bytes to dst.
        // It returns a new byte slice that is the result of the append operation.
	AppendDecompressed(dst, src []byte, decompressedLength uint32) ([]byte, error)
}
```

The lz4 library does not support append like semantics so we're going to have to expand `dst` slice before passing it to the lz4 library functions. If `cap(dst)` is not large enough then we'll have to allocate a new byte slice and append it to `dst` instead.

I used the name `Compress` and `Decompress` instead of `Encode` / `Decode` because it felt more natural to me but I'm fine with keeping `Encode` and `Decode`.

Reply:
Could you please look at [this](https://github.com/worryg0d/gocql/commit/53c721ff91de86f7f02f16eeeefc4a15a290ad22)? I tried to implement this append-like API, but I'm unsure if I should implement the `AppendCompressed` and `AppendDecompressed` methods for our snappy wrapper...

> I used the name Compress and Decompress instead of Encode / Decode because it felt more natural to me but I'm fine with keeping Encode and Decode. Same for WithLength/WithSize

I like your variants. If nobody has anything against then I'll use them.

Reply:
> I'm unsure if I should implement the AppendCompressed and AppendDecompressed methods for our snappy wrapper...

I think it's fine to make them panic since protocol v5 doesn't support snappy anyway.

Reply:
I added the suggested by @joao-r-reis solution. @martin-sucha Could you please take a look at [this](https://github.com/apache/cassandra-gocql-driver/pull/1822/commits/86130c0836b49d3ad253725f8d9e837443afd7bd)?